design/physical-network-api-prototype.md

```diff
@@ -0,0 +1,176 @@
+         * [Test Plan](#test-plan)
+         * [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+         * [Version Skew Strategy](#version-skew-strategy)
+      * [Drawbacks [optional]](#drawbacks-optional)
+      * [Alternatives [optional]](#alternatives-optional)
+      * [References](#references)
+
+<!-- Added by: stack, at: 2019-02-15T11:41-05:00 -->
+
+<!--te-->
+
+## Summary
+
+The Metal³ project is currently centered around an API for managing physical
+hosts.  This proposal is to explore expanding the scope to also include an API
+for managing physical network devices.  This exploration would be done by
+starting with a prototype that can configure some aspects of a top-of-rack
+(ToR) switch.
+
+## Motivation
+
+Metal³ follows the paradigm of Kubernetes Native Infrastructure (KNI), which is
+an approach to use Kubernetes to manage underlying infrastructure.  Managing
+the configuration of some physical network devices is closely related to
+managing physical hosts.
+
+As bare metal hosts are provisioned or later repurposed, there may be
+corresponding physical network changes that must be made, such as reconfiguring
+a ToR switch port.  If the provisioning of the physical host is managed through
+a Kubernetes API, it would be convenient to be able to reconfigure related
+network devices using a similar API.
+
+### Goals
+
+Produce a `physical-network-api` prototype, which includes the following:
+
+* A `Switch` CRD with one or more fields that corresponding to switch
+  configuration, ideally with a vendor neutral definition
+* A demonstration of using this prototype API to configure a switch
+* A documented retrospective that includes lessons learned and proposed next
+  steps
+* Evaluate the re-use of a single existing network device configuration
+  technology
+
+### Non-Goals
+
+Out of scope:
+
+* Testing with anything more than a single switch model
+* Prototyping an API for anything beyond a Switch
+* Full configuration possibilities of a switch
+* Evaluation of all potential technologies that could be re-used for network
+  device configuration.  Future prototypes can be done to consider other
+  alternatives if desired.
+
+These items may become in scope after a prototype is reviewed and next steps
+discussed.
+
+## Proposal
+
+The prototype should explore the re-use of [Ansible
+Networking](https://docs.ansible.com/ansible/latest/network/index.html) to
+perform device configuration.  Ansible includes modules for managing the
+configuration of many different network devices.  Creation of CRDs in this new
+API can use modeling from Ansible as inspiration, particularly if there is any
+vendor neutral modeling already done there.
+
+Another project that could be used as inspiration is the [Ansible Networking
+Neutron ML2 Driver](https://networking-ansible.readthedocs.io/en/latest/) which
+created a Python API to abstract generic switch configuration on top of Anisble
+networking modules.
```

Comment:
Note that the relevant part of this driver was split out into https://github.com/ansible-network/network-runner.

The capabilities exposed by the API at the moment are confined to creating and deleting VLANs, and assigning them to switch ports (either as access ports or using 802.1Q VLAN trunking). So far there's no L3 support (adding IP addresses to VLANs, setting up routes, configuring DHCP Proxying, &c.), although in principle anything that's supported by ansible networking should be easy enough to add.

Reply:
Thanks for the tip.  I'll add this to the doc as a reference.